File: Lua/cache.lua
local cache = {}

function cache:load(urls)
  self._urls = {}
  for _, u in ipairs(urls) do
    self._urls[u.id] = { original = u.original, accesses = tonumber(u.accesses) or 0 }
  end
end

function cache:get_all()
  local result = {}
  for k, v in pairs(self._urls) do
    result[k] = { original = v.original, accesses = v.accesses }
  end
  return result
end

function cache:get(id)
  local v = self._urls[id]
  if v then
    return { original = v.original, accesses = v.accesses }
  end
  return nil
end

function cache:exists(id)
  return self._urls[id] ~= nil
end

function cache:set(id, info)
  self._urls[id] = info
end

function cache:delete(id)
  self._urls[id] = nil
end

function cache:increment_accesses(id)
  local v = self._urls[id]
  if v then
    v.accesses = v.accesses + 1
    return v.accesses
  end
  return 0
end

return cache

File: Rust/src/cache.rs
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;

use crate::models::{Url, UrlInfo};

#[derive(Clone)]
pub struct UrlCache {
    inner: Arc<RwLock<HashMap<String, UrlInfo>>>,
}

impl UrlCache {
    pub fn new() -> Self {
        Self {
            inner: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub async fn load(&self, urls: Vec<Url>) {
        let mut map = self.inner.write().await;
        map.clear();
        for u in urls {
            map.insert(u.id, UrlInfo { original: u.original, accesses: u.accesses });
        }
    }

    pub async fn get_all(&self) -> HashMap<String, UrlInfo> {
        self.inner.read().await.clone()
    }

    pub async fn get(&self, id: &str) -> Option<UrlInfo> {
        self.inner.read().await.get(id).cloned()
    }

    pub async fn exists(&self, id: &str) -> bool {
        self.inner.read().await.contains_key(id)
    }

    pub async fn set(&self, id: String, info: UrlInfo) {
        self.inner.write().await.insert(id, info);
    }

    pub async fn delete(&self, id: &str) {
        self.inner.write().await.remove(id);
    }

    pub async fn increment_accesses(&self, id: &str) -> i64 {
        let mut map = self.inner.write().await;
        if let Some(info) = map.get_mut(id) {
            info.accesses += 1;
            info.accesses
        } else {
            0
        }
    }
}
